modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java

```diff
@@ -0,0 +1,1149 @@
+            return boolean.class;
+        }
+        
+        // ordinary numeric promotion
+        if (a == double.class || b == double.class) {
+            return double.class;
+        } else if (a == float.class || b == float.class) {
+            return float.class;
+        } else if (a == long.class || b == long.class) {
+            return long.class;
+        } else {
+            return int.class;
+        }
+    }
+    
+    private static final Lookup PRIV_LOOKUP = MethodHandles.lookup();
+
+    private static final Map<Class<?>,Map<String,MethodHandle>> TYPE_OP_MAPPING = Collections.unmodifiableMap(
+        Stream.of(boolean.class, int.class, long.class, float.class, double.class, Object.class)
+            .collect(Collectors.toMap(Function.identity(), type -> {
+                try {
+                    Map<String,MethodHandle> map = new HashMap<>();
+                    MethodType unary = MethodType.methodType(type, type);
+                    MethodType binary = MethodType.methodType(type, type, type);
+                    MethodType comparison = MethodType.methodType(boolean.class, type, type);
+                    MethodType shift = MethodType.methodType(type, type, long.class);
+                    Class<?> clazz = PRIV_LOOKUP.lookupClass();
+                    map.put("not",   PRIV_LOOKUP.findStatic(clazz, "not",  unary));
+                    map.put("neg",   PRIV_LOOKUP.findStatic(clazz, "neg",  unary));
+                    map.put("plus",  PRIV_LOOKUP.findStatic(clazz, "plus", unary));
+                    map.put("mul",   PRIV_LOOKUP.findStatic(clazz, "mul",  binary));
+                    map.put("div",   PRIV_LOOKUP.findStatic(clazz, "div",  binary));
+                    map.put("rem",   PRIV_LOOKUP.findStatic(clazz, "rem",  binary));
+                    map.put("add",   PRIV_LOOKUP.findStatic(clazz, "add",  binary));
+                    map.put("sub",   PRIV_LOOKUP.findStatic(clazz, "sub",  binary));
+                    map.put("and",   PRIV_LOOKUP.findStatic(clazz, "and",  binary));
+                    map.put("or",    PRIV_LOOKUP.findStatic(clazz, "or",   binary));
+                    map.put("xor",   PRIV_LOOKUP.findStatic(clazz, "xor",  binary));
+                    map.put("eq",    PRIV_LOOKUP.findStatic(clazz, "eq",   comparison));
+                    map.put("lt",    PRIV_LOOKUP.findStatic(clazz, "lt",   comparison));
+                    map.put("lte",   PRIV_LOOKUP.findStatic(clazz, "lte",  comparison));
+                    map.put("gt",    PRIV_LOOKUP.findStatic(clazz, "gt",   comparison));
+                    map.put("gte",   PRIV_LOOKUP.findStatic(clazz, "gte",  comparison));
+                    map.put("lsh",   PRIV_LOOKUP.findStatic(clazz, "lsh",  shift));
+                    map.put("rsh",   PRIV_LOOKUP.findStatic(clazz, "rsh",  shift));
+                    map.put("ush",   PRIV_LOOKUP.findStatic(clazz, "ush",  shift));
+                    return map;
```

Comment:
I think this map should also be unmodifiable like the outer one. It is not so important, because the whole thing is private, but for consistency. Alternatively remove the unmodifiable from outer map, too.